pkg/parser/parser.go

```diff
@@ -647,52 +656,56 @@ func (p *Parser) parseForStatement(scope *scope) Node {
 }
 
 func (p *Parser) parseWhileStatement(scope *scope) Node {
-	tok := p.cur
+	result := &While{}
+	result.Token = p.cur
 	p.advance() // advance past WHILE token
-	scope = newInnerScope(scope)
-	condition := p.parseCondition(scope)
+	scope = scope.innerScope(result)
+	result.Condition = p.parseCondition(scope)
 	p.advancePastNL()
-	block := p.parseBlock(scope)
+	result.Block = p.parseBlock(scope)
 	p.assertEnd()
 	p.advancePastNL()
-	return &While{
-		ConditionalBlock{
-			Token:     tok,
-			Condition: condition,
-			Block:     block,
-		},
+	return result
+}
+
+func inLoop(s *scope) bool {
+	for ; s != nil; s = s.outer {
+		if _, ok := s.block.(*While); ok {
+			return true
+		}
 	}
+	return false
 }
 
 func (p *Parser) parseIfStatement(scope *scope) Node {
-	ifStmt := &If{Token: p.cur}
-	ifStmt.IfBlock = p.parseIfConditionalBlock(scope)
+	result := &If{Token: p.cur}
+	result.IfBlock = p.parseIfConditionalBlock(result, scope)
 	// else if blocks
 	for p.cur.TokenType() == lexer.ELSE && p.peek.TokenType() == lexer.IF {
 		p.advance() // advance past ELSE token
-		elseIfBlock := p.parseIfConditionalBlock(scope)
-		ifStmt.ElseIfBlocks = append(ifStmt.ElseIfBlocks, elseIfBlock)
+		elseIfBlock := p.parseIfConditionalBlock(result, scope)
+		result.ElseIfBlocks = append(result.ElseIfBlocks, elseIfBlock)
 	}
 	// else block
 	if p.cur.TokenType() == lexer.ELSE {
 		p.advance() // advance past ELSE token
 		p.assertEOL()
 		p.advancePastNL()
-		ifStmt.Else = p.parseBlock(newInnerScope(scope))
+		result.Else = p.parseBlock(scope.innerScope(result))
 	}
 	p.assertEnd()
 	p.advancePastNL()
-	return ifStmt
+	return result
 }
 
-func (p *Parser) parseIfConditionalBlock(scope *scope) *ConditionalBlock {
-	tok := p.cur
+func (p *Parser) parseIfConditionalBlock(ifStmt *If, scope *scope) *ConditionalBlock {
+	result := &ConditionalBlock{Token: p.cur}
 	p.advance() // advance past IF token
-	scope = newInnerScope(scope)
-	condition := p.parseCondition(scope)
+	scope = scope.innerScope(ifStmt)
+	result.Condition = p.parseCondition(scope)
```

Comment:
the scope of the condition should be the outer scope, not the inner one. it cannot refer to variables declared inside the block. it may not technically matter, because at this point the inner scope is empty and the condition cannot declare vars (although if it could, should they be in the scope of the block or half-way?an extra hidden middle scope? :pineapple: on the question)

Reply:
ok, fixed. thanks.

Reply:
hmm, I thought about this some more:
A statement has never access to the variables declared "later" in the same block - so in that sense there is no difference between the scope of the condition and the scope of the following block.

However, for a `for` loop the "condition" is what is going on inside the `for i := range 10` statement. i will be available inside the loop but not outside the loop, so from my perspective i is defined in the same block as the loop.